sdk/include/platform/sunburst/platform-usbdev.hh

```diff
@@ -0,0 +1,633 @@
+	                                 bool     &isSetup,
+	                                 uint32_t *destination) volatile
+	{
+		if (!(usbStatus & static_cast<uint32_t>(UsbStatusField::ReceiveDepth)))
+		{
+			return -1; // No packets received
+		}
+
+		uint32_t received = receiveBuffer;
+
+		typedef ReceiveBufferField Reg;
+		endpointId = (received & static_cast<uint32_t>(Reg::EndpointId)) >> 20;
+		size       = (received & static_cast<uint32_t>(Reg::Size)) >> 8;
+		isSetup    = (received & static_cast<uint32_t>(Reg::Setup)) != 0;
+		bufferId   = (received & static_cast<uint32_t>(Reg::BufferId)) >> 0;
+
+		// Reception of Zero Length Packets occurs in the Status Stage of IN
+		// Control Transfers.
+		if (size > 0)
+		{
+			usbdev_transfer(destination, buffers(bufferId), size, false);
+		}
+		return 0;
+	}
+
+	private:
+	/**
+	 * Return a pointer to the given offset within the USB device register
+	 * space; this is used to access the packet buffer memory.
+	 *
+	 * @param bufferId The buffer number to access the packet buffer memory for
+	 *
+	 * @returns A volatile pointer to the buffer's memory.
+	 */
+	volatile uint32_t *buffers(uint8_t bufferId) volatile
+	{
+		const uint32_t Offset = BufferStartAddress + bufferId * MaxPacketLength;
+		const uintptr_t Address = reinterpret_cast<uintptr_t>(this) + Offset;
+		return reinterpret_cast<uint32_t *>(Address);
+	}
+
+	/**
+	 * Perform a transfer to or from packet buffer memory. This function is
+	 * hand-optimised to perform a faster, unrolled, word-based data transfer
+	 * for efficiency.
+	 *
+	 * @param destination A pointer to transfer the source data to.
+	 * @param source A pointer to the data to be transferred.
+	 * @param size The size of the data pointed to by `source`.
+	 * @param toDevice True if the transfer is to the device (e.g. when sending
+	 * a packet), and False if not (e.g. when receiving a packet).
+	 */
+	static void usbdev_transfer(volatile uint32_t       *destination,
+	                            const volatile uint32_t *source,
+	                            uint8_t                  size,
+	                            bool                     toDevice)
+	{
+		// Unroll word transfer. Each word transfer is 4 bytes, so we must round
+		// to the closest multiple of (4 * words) when unrolling.
+		constexpr uint8_t  UnrollFactor = 4u;
+		constexpr uint32_t UnrollMask   = (UnrollFactor * 4u) - 1;
+
+		// Round down to the previous multiple for unrolling
+		const uint32_t  UnrollSize = (size & ~UnrollMask);
+		const uint32_t *sourceEnd  = reinterpret_cast<uint32_t *>(
+          reinterpret_cast<uintptr_t>(source) + UnrollSize);
+
+		// Unrolled to mitigate loop overheads.
+		// Ensure the unrolling here matches `UnrollFactor`.
+		while (source < sourceEnd)
+		{
+			destination[0] = source[0];
+			destination[1] = source[1];
+			destination[2] = source[2];
+			destination[3] = source[3];
+			destination += UnrollFactor;
+			source += UnrollFactor;
+		}
```

Comment:
I don't really like the manual loop unrolling here.  Is there a reason that this isn't just a memcpy?  The memcpy loop will do each copy as a capability, so even without unrolling it will be a fairly similar speed to this.

With the current compiler, this structure actually hits one of the pathological cases for missed optimisation and so I'd expect it to be much slower.

Reply:
I've added a comment explaining why this was done. Does it explain it well? Are you happy with the reasoning?

Reply:
I worry that the compiler will reroll and unroll this.  It's fragile to do it like this because, from a C++ abstract machine perspective, it's no different than memcpy.